File: update_nodes.py
import json
from prisma import Prisma
import asyncio

# Function to remove duplicate nodes
def remove_duplicates(data):
    seen_identifiers = set()
    unique_data = []
    for node in data:
        identifier = node['identifier']
        if identifier not in seen_identifiers:
            seen_identifiers.add(identifier)
            unique_data.append(node)
    return unique_data

# Function to add restVersion if it is missing
def add_restVersion(data):
    for node in data:
        if 'restVersion' not in node:
            node['restVersion'] = 'v4'
    return data

# Load data from nodes.json
with open('nodes.json') as f:
    data = json.load(f)

# Add restVersion if it is missing
data = add_restVersion(data)

# Remove duplicates
unique_data = remove_duplicates(data)

# Check if any nodes were removed
if len(unique_data) < len(data):
    print("Duplicate nodes removed.")

    # Save updated data back to nodes.json
    with open('nodes.json', 'w') as f:
        json.dump(unique_data, f, indent=4)
else:
    print("No duplicate nodes found.")

# Save data with restVersion added
with open('nodes.json', 'w') as f:
    json.dump(unique_data, f, indent=4)

# Update database with unique nodes
async def update_nodes(nodes):
    print("Updating nodes in database...")
    prisma = Prisma()
    await prisma.connect()
    # delete all nodes
    await prisma.nodes.delete_many()
    # create new nodes
    await prisma.nodes.create({
        "nodes": nodes
    })
    await prisma.disconnect()

json_nodes = json.dumps(unique_data)
# Call updateNodes with unique_data
asyncio.run(update_nodes(json_nodes))

File: update_nodes.ts
import * as fs from 'fs';
import { PrismaClient } from '@prisma/client';
import { PrismaLibSQL } from "@prisma/adapter-libsql";
import { createClient } from "@libsql/client";

const libsql = createClient({
    url: process.env.TURSO_DATABASE_URL || '',
    authToken: process.env.TURSO_AUTH_TOKEN,
});

const adapter = new PrismaLibSQL(libsql);
// Define types for node objects
interface Node {
    host: string;
    identifier: string;
    password?: string;
    port?: number;
    restVersion?: string;
    secure?: boolean;
    authorId?: string;
}

// Function to remove duplicate nodes
function removeDuplicates(data: Node[]): Node[] {
    const seenIdentifiers = new Set<string>();
    const uniqueData: Node[] = [];
    for (const node of data) {
        const { identifier } = node;
        if (!seenIdentifiers.has(identifier)) {
            seenIdentifiers.add(identifier);
            uniqueData.push(node);
        }
    }
    return uniqueData;
}

// Function to add restVersion if it is missing
function addRestVersion(data: Node[]): Node[] {
    for (const node of data) {
        if (!node.restVersion) {
            node.restVersion = 'v4';
        }
    }
    return data;
}

// Load data from nodes.json
const data: Node[] = JSON.parse(fs.readFileSync('nodes.json', 'utf8'));

// Add restVersion if it is missing
const dataWithRestVersion = addRestVersion(data);

// Remove duplicates
const uniqueData = removeDuplicates(dataWithRestVersion);

// Check if any nodes were removed
if (uniqueData.length < data.length) {
    console.log("Duplicate nodes removed.");

    // Save updated data back to nodes.json
    fs.writeFileSync('nodes.json', JSON.stringify(uniqueData, null, 4));
} else {
    console.log("No duplicate nodes found.");
}

// Update database with unique nodes
async function updateNodes(nodes: Node[]): Promise<void> {
    console.log("Updating nodes in database...");
    const prisma = new PrismaClient({ adapter });
    await prisma.$connect();
    // delete all nodes
    await prisma.nodes.deleteMany();
    // create new nodes

    await prisma.nodes.createMany({
        data: {
            nodes: JSON.stringify(nodes)
        }
    });
    await prisma.$disconnect();
}

// Call updateNodes with uniqueData
updateNodes(uniqueData)
    .then(() => console.log("Nodes updated successfully."))
    .catch(err => console.error("Error updating nodes:", err));
